File: src/backend/controllers/ChatController.ts
import { Request, Response, Router } from "express";
import { ChatService } from "../services/ChatService";
import { userExtractor } from "./UserController";

export class ChatController {
  private chatService = new ChatService();
  public router = Router();

  constructor() {
    this.router.get("/", userExtractor, this.getHistory.bind(this));
    this.router.post("/send", userExtractor, this.sendMessage.bind(this));
    this.router.post("/clear", userExtractor, this.clearChat.bind(this));
    this.router.post("/tts", userExtractor, this.generateTTS.bind(this));
  }

  private getHistory(req: Request, res: Response): void {
    try {
      const courseId = req.query.courseId ? Number(req.query.courseId) : undefined;
      const history = this.chatService.getChatHistory(req.userId!, courseId);
      res.json(history);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private async sendMessage(req: Request, res: Response): Promise<void> {
    try {
      const { message, courseId, aiMode } = req.body;
      if (!message) {
        res.status(400).json({ error: "Message content is required" });
        return;
      }
      const result = await this.chatService.sendMessage(
        req.userId!,
        courseId ? Number(courseId) : undefined,
        message,
        aiMode || "Text"
      );
      res.json(result);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private clearChat(req: Request, res: Response): void {
    try {
      this.chatService.clearChat(req.userId!);
      res.json({ success: true, message: "Chat history cleared" });
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private async generateTTS(req: Request, res: Response): Promise<void> {
    try {
      const { text } = req.body;
      if (!text) {
        res.status(400).json({ error: "Text content is required for TTS" });
        return;
      }
      const audioBase64 = await this.chatService.generateTTS(text, req.userId!);
      res.json({ audioBase64 });
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }
}

File: converted_code/backend/src/main/java/com/zeriq/ai/controller/ChatController.java
package com.zeriq.ai.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.zeriq.ai.model.Conversation;
import com.zeriq.ai.service.ChatService;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/chat")
@CrossOrigin(origins = "*")
public class ChatController {

    @Autowired
    private ChatService chatService;

    @GetMapping
    public ResponseEntity<?> getChatHistory(@RequestHeader(value = "x-user-id", required = false) Long userId) {
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Unauthorized. Missing User ID header."));
        }
        List<Conversation> history = chatService.getHistory(userId);
        return ResponseEntity.ok(history);
    }

    @PostMapping("/send")
    public ResponseEntity<?> sendMessage(
            @RequestHeader(value = "x-user-id", required = false) Long userId,
            @RequestBody Map<String, Object> payload) {
        
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Unauthorized. Missing User ID header."));
        }

        try {
            String message = payload.get("message").toString();
            Long courseId = payload.get("courseId") != null ? Long.valueOf(payload.get("courseId").toString()) : null;
            String aiMode = payload.get("aiMode") != null ? payload.get("aiMode").toString() : "Text";

            Conversation aiResponse = chatService.generateResponse(userId, courseId, message, aiMode);
            
            // Returns both the text response and optional audio Base64 payload
            return ResponseEntity.ok(Map.of(
                    "response", aiResponse,
                    "audioBase64", aiResponse.getAudioBase64() != null ? aiResponse.getAudioBase64() : ""
            ));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("error", "Invalid body parameters: " + e.getMessage()));
        }
    }

    @PostMapping("/tts")
    public ResponseEntity<?> generateTTS(@RequestBody Map<String, String> payload) {
        String text = payload.get("text");
        if (text == null || text.trim().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("error", "Text is required for TTS synthesis."));
        }

        // Returns empty or mock audio string. Frontend fallbacks to Web Speech synthesizers seamlessly.
        return ResponseEntity.ok(Map.of("audioBase64", ""));
    }

    @PostMapping("/clear")
    public ResponseEntity<?> clearChat(@RequestHeader(value = "x-user-id", required = false) Long userId) {
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Unauthorized. Missing User ID header."));
        }
        
        chatService.clearHistory(userId);
        return ResponseEntity.ok(Map.of("success", true, "message", "Chat history cleared"));
    }
}
